### ISSUE TITLE
BGRU Model Architecture

### ISSUE THREAD

**Issue by pravanjang** [pravanjang/stock-prediction-system#5]
**Code to Generate:**
Complete script `models/bgru_base.py` that implements Bidirectional GRU:

**Required Class:**
```python
class BGRUPredictor:
    def __init__(self, input_dim=5, hidden_dim=128, num_layers=2, dropout=0.3):
        """
        Bidirectional GRU Architecture:
        - Input: OHLCV sequences (shape: [batch, sequence_length, 5])
        - BGRU Layer 1: 128 units, bidirectional, dropout=0.3
        - BGRU Layer 2: 64 units, bidirectional, dropout=0.3
        - Dense Layer: 32 units, ReLU activation
        - Output Layer: 1 unit, Sigmoid activation (binary classification)
        """
        
    def build_model(self):
        """Constructs the BGRU model using PyTorch"""
        
    def prepare_sequences(self, df, sequence_length=60):
        """
        Prepares sequential data for BGRU:
        - Creates sliding windows of 60 time steps (15 hours of data)
        - Normalizes using rolling z-score (window=100)
        - Returns: X (sequences), y (targets)
        """
        
    def normalize_data(self, df, method='rolling_zscore', window=100):
        """Normalizes OHLCV data using rolling z-score"""
        
    def train(self, train_df, val_df, epochs=50, batch_size=64, lr=0.001):
        """
        Training configuration:
        - Optimizer: Adam with lr=0.001
        - Loss: Binary Cross-Entropy
        - Batch size: 64
        - Epochs: 50 with early stopping (patience=10)
        - Learning rate scheduler: ReduceLROnPlateau (factor=0.5, patience=5)
        """
        
    def predict(self, test_df):
        """Generate predictions on test data"""
        
    def save_model(self, path='models/checkpoints/bgru_baseline.pt'):
        """Save model checkpoint"""
        
    def load_model(self, path):
        """Load model checkpoint"""
```

**Model Architecture Details:**
```
Input Layer: [batch, 60, 5] (60 time steps, 5 features: OHLCV)
    ↓
BGRU Layer 1: 128 units (256 total: 128 forward + 128 backward)
    ↓
Dropout: 0.3
    ↓
BGRU Layer 2: 64 units (128 total: 64 forward + 64 backward)
    ↓
Dropout: 0.3
    ↓
Dense Layer: 32 units, ReLU
    ↓
Dropout: 0.2
    ↓
Output Layer: 1 unit, Sigmoid
```

**Training Configuration:**
- Optimizer: Adam (lr=0.001, weight_decay=1e-5)
- Loss function: Binary Cross-Entropy with optional class weights
- Batch size: 64 (adjust based on GPU memory)
- Epochs: 50 with early stopping (patience=10)
- Learning rate scheduler: ReduceLROnPlateau
- Gradient clipping: max_norm=1.0 (prevents exploding gradients)

**Model Checkpointing:**
- Auto-save best model based on validation accuracy to `models/checkpoints/`
- Save training history to `models/checkpoints/training_history.json`
- Log training curves to `models/checkpoints/training_curves.png`

**Execution:**
```bash
python models/bgru_base.py --train --data_dir data/processed/ --sequence_length 60
```

**Logging Requirements:**
- Print epoch-wise metrics (loss, accuracy, validation accuracy) to console
- Save detailed logs to `models/logs/training.log`
- Create progress bar for training (use tqdm)
- Plot training/validation curves after training
